Remove `-ptr` from the help output of `service create`

yongtang opened the pull request:
This fix is based on the comment:
https://github.com/docker/docker/pull/28147#discussion_r86996347

Previously the output string of the `DurationOpt` is `duration-ptr` and `Uint64Opt` is `uint64-ptr`. While it is clear to developers, for a normal user `-ptr` might not be very informative.

On the other hand, the default value of `DurationOpt` and `Uint64Opt` has already been quite informative: `none`. That means if no flag provided, the value will be treated as none. (a ptr with nil as the default)

For that reason this fix removes the `-ptr`.

Also, the output in the docs of `service create` has been quite out-of-sync with the true output. So this fix updates the docs to have the most up-to-date help output of `service create --help`.

This fix is related to #28147.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Remove `-ptr` from the help output of `service create`

This fix is based on the comment:
https://github.com/docker/docker/pull/28147#discussion_r86996347

Previously the output string of the `DurationOpt` is `duration-ptr`
and `Uint64Opt` is `uint64-ptr`. While it is clear to developers,
for a normal user `-ptr` might not be very informative.

On the other hand, the default value of `DurationOpt` and `Uint64Opt`
has already been quite informative: `none`. That means if no flag
provided, the value will be treated as none.
(like a ptr with nil as the default)

For that reason this fix removes the `-ptr`.

Also, the output in the docs of `service create` has been quite
out-of-sync with the true output. So this fix updates the docs
to have the most up-to-date help output of `service create --help`.

This fix is related to #28147.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

vdemeester reviewed (approved):
LGTM 🐸
/cc @dnephin @thaJeztah

thaJeztah commented on docs/reference/commandline/service_create.md:
```diff
@@ -21,46 +21,46 @@ Usage:  docker service create [OPTIONS] IMAGE [COMMAND] [ARG...]
+  -p, --publish list                       Publish a port as a node port (default [])
+      --replicas uint                      Number of tasks (default none)
+      --reserve-cpu NanoCPUs               Reserve CPUs (default 0.000)
+      --reserve-memory MemoryBytes         Reserve Memory (default 0 B)
```
Some of the others could also be improved; do you want to do that in this PR (if possible)?

For example;

- `MemoryBytes` could possibly be simply `bytes` https://github.com/docker/docker/blob/69efb4652c1f619f64148fd21971ff18ac575899/cli/command/service/opts.go#L35
- `stringSlice`, would that be the same as "`list`" from a user perspective? Or "`string`"?

yongtang replied:
@thaJeztah The `stringSlice` is similar to `list`. Replacing  `[] string` (`StringSliceVar`) to `opts.ListOpts` will show up as `list`. Let me update the PR shortly.

thaJeztah replied:
Discussing with @dnephin;

We can use

- `bytes` for `MemoryBytes`
- `list` for `stringSlice`
- `decimal` for `NanoCPUs`

I think those are the important ones (`mount` may be ok to keep, because it is quite a detailed format)

thaJeztah commented on cli/command/service/opts.go:
```diff
@@ -101,9 +101,9 @@ func (i *Uint64Opt) Set(s string) error {
+// Type returns the type of this option, which will be displayed in `--help` output
 func (i *Uint64Opt) Type() string {
-	return "uint64-ptr"
+	return "uint64"
```
I wonder if `uint64` is too much information already; `integer` perhaps (if we give a suitable error message for negative values)

Open to suggestions though /cc @dnephin @bfirsh @vdemeester

thaJeztah replied:
I see other software even using simply `number`

vdemeester replied:
Hum, I like the simplicity of `number` 👼

mstanleyjones replied:
As long as nobody would try to use a floating-point...

thaJeztah replied:
Yes, we may want to come up with some standard terms if we need to differentiate; suggestions welcome :smile:

thaJeztah replied:
@yongtang ^^

justincormack replied:
number is much better.

thaJeztah commented:
Adding this to the 1.13 milestone, as this changed since 1.12

vieux commented:
@thaJeztah should we wait for #28147 before merging ?

thaJeztah commented:
@vieux yes, let's wait for https://github.com/docker/docker/pull/28147 - there still some discussion on this one, and changes are small, so we should be able to get those in before code freeze

vieux commented:
#28147 is merged @thaJeztah

yongtang commented:
Made an update but didn't see some of the discussions before push (`decimal for NanoCPUs`). Will update the PR again shortly.

yongtang commented:
Thanks all for the suggestions and review.

Updated again. Here is the list of changes so far:
* `bytes` for `MemoryBytes`
* `list` for `stringSlice`
* `decimal` for `NanoCPUs`
* `duration` for `duration-ptr`
* *`uint` for `uint64-ptr`* (Do we want `number` instead?)

Also, there is still a `float32` in the output:
```
--update-max-failure-ratio float32   Failure rate to tolerate during an update
```
I am thinking maybe we could change it to `float` (this is the only one with the `32`)?

yongtang commented:
Rebased and updated the PR. Now the complete list of changes is:
* `bytes` for `MemoryBytes`
* `list` for `stringSlice`
* `decimal` for `NanoCPUs`
* `duration` for `duration-ptr`
* `uint` for `uint64-ptr` (Do we want `number` instead?)
* `float` for `float32` (Let me know if we want to keep `float32` so that I could remove it)

thaJeztah reviewed (approved):
LGTM

/cc @justincormack

yongtang commented:
Rebased to resolve merge conflict.

thaJeztah commented:
oh, hmf, looks like I just merged a PR that conflicted with this, so sorry @yongtang, it needs another rebase 😞

yongtang commented:
Thanks @thaJeztah. The PR has been rebased.

vdemeester reviewed (approved):
LGTM 🐸
/cc @dnephin

dnephin commented on cli/command/service/opts.go:
```diff
@@ -119,12 +119,32 @@ func (i *Uint64Opt) Value() *uint64 {
+
+func (f *floatValue) Value() float32 {
+	return float32(*f)
+}
```
I would prefer that we don't create a bunch of these just to modify the `Type()` , but one is fine I guess.

dnephin reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM, thanks!
